torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -67,21 +80,36 @@ def __init__(self):
         self._training_state: TrainingState = TrainingState.IDLE
         self._states_to_forward_prefetch: List[FSDPState] = []
         self._states_to_backward_prefetch: List[FSDPState] = []
+        self._modules_to_run_forward: Set[nn.Module] = set()
 
     # Define a separate init since `__init__` is called in the contract
     def init(
-        self, module: nn.Module, device: torch.device, mp_policy: MixedPrecisionPolicy
+        self,
+        modules: Tuple[nn.Module, ...],
+        device: torch.device,
+        mp_policy: MixedPrecisionPolicy,
     ) -> None:
-        _insert_module_state(module, self)
-        self._module = module
+        for module in modules:
+            _insert_module_state(module, self)
+        self._modules = modules
         self._device = device
         self._mp_policy = mp_policy
-        self._pre_forward_hook_handle = module.register_forward_pre_hook(
-            self._pre_forward, prepend=True, with_kwargs=True
-        )
-        self._post_forward_hook_handle = module.register_forward_hook(
-            self._post_forward, prepend=False
-        )
+        if len(modules) == 1:
+            self._pre_forward_hook_handle = modules[0].register_forward_pre_hook(
+                self._pre_forward, prepend=True, with_kwargs=True
+            )
+            self._post_forward_hook_handle = modules[0].register_forward_hook(
+                self._post_forward, prepend=False
+            )
+        else:
```

Comment:
The main changes are (1) generalizing internally from `nn.Module` to `Tuple[nn.Module, ...]` and (2) using these newly-defined forward hooks if taking the `List[nn.Module]` path.